### PR TITLE
`tsx` syntax support

### LINKED ISSUE #469: Feature Request : Support JSX
@sebilasse commented on [Fri Apr 28 2017](https://github.com/dojo/cli/issues/144)

Wondered how my "html preferring" users could write dojo2 widgets in a declarative way and just read https://medium.com/maquette-news/maquette-2-2-now-supports-jsx-7da7b0a1dc98
This is so nice. I could think of a great benefit for dojo users.

---

@dylans commented on [Fri Apr 28 2017](https://github.com/dojo/cli/issues/144#issuecomment-298147394)

We have support for this in a branch on one of our contributor's repos. While we think jsx is a step in the wrong direction compared to hyperscript with regards to functional reactive programming, and we're trying to avoid the temptation of supporting too many different ways of doing things (which can lead to more complexity), we will reconsider adding this as option before the release of Dojo 2.

---

@dylans commented on [Sat Apr 29 2017](https://github.com/dojo/cli/issues/144#issuecomment-298166412)

I'm going to move this issue to widget-core since that is where it would be supported

### PR DESCRIPTION
**Pull request by agubler**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code matches the [style guide](https://github.com/dojo/meta/blob/master/STYLE.md)
* [x] Unit or Functional tests are included in the PR

**Description:**

Adds support for using `tsx` syntax to author Dojo 2 widgets within the `render` function.

```tsx
return (
	<footer classes={this.classes(css.footer)}>
		<span classes={this.classes(css.todoCount)}>
			<strong>{`${activeCount}`}</strong>
			<span>{`${count}`}</span>
		</span>
		<TodoFilter activeFilter={activeFilter} />
		{ completedItems ? ( <button onclick={this.clearCompleted} /> ) : ( null ) }
	</footer>
);
```

Props to @matt-gadd for the `tsx` pragma and function implementation (the important parts!).

Resolves #469

### CODE REVIEW

**Comment by agubler** [comment]
@rorticus added some blurb on JSX in the README.

File: README.md [hunk 1]
```diff
@@ -173,6 +174,51 @@ w('my-widget', properties, children);
 The example above that uses a string for the `widgetConstructor `, is taking advantage of the [widget registry](#widget-registry) functionality.
 The widget registry allows for the lazy loading of widgets.
 
+### tsx
+
+In additional to the programatic functions `v` and `w`, widget-core supports the use of the `jsx` syntax known as [`tsx`](https://www.typescriptlang.org/docs/handbook/jsx.html) in TypeScript.
+
+To start to use `jsx` in your project the widgets need to be named with a `.tsx` extension and some configuration is required in the projects `tsconfig.json`:
```
Comment (dylans) [inline README.md]: projects -> project's

File: README.md [hunk 2]
```diff
@@ -173,6 +174,51 @@ w('my-widget', properties, children);
...
+				</span>
+				<TodoFilter activeFilter={activeFilter} />
+				{ completed ? ( <button onclick={clear} /> ) : ( null ) }
+			</footer>
+		);
+	}
+}
+```
+
+**Note:** Unfortunately `tsx` is not directly used so needs to be ignored by linters.
```
Comment (dylans) [inline README.md]: Not directly used with what? This could be clearer

File: README.md [hunk 3]
```diff
@@ -173,6 +174,51 @@ w('my-widget', properties, children);
 The example above that uses a string for the `widgetConstructor `, is taking advantage of the [widget registry](#widget-registry) functionality.
 The widget registry allows for the lazy loading of widgets.
 
+### tsx
+
+In additional to the programatic functions `v` and `w`, widget-core supports the use of the `jsx` syntax known as [`tsx`](https://www.typescriptlang.org/docs/handbook/jsx.html) in TypeScript.
```
Comment (dylans) [inline README.md]: To be slightly more explicit that this is an alternative approach:

widget-core supports the use -> widget-core optionally supports the use

File: tests/unit/tsx.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import { HNODE } from './../../src/d';
+
+registerSuite({
+	name: 'tsx',
+	'create a registry wrapper'() {
+		const RegistryWrapper = fromRegistry<WidgetProperties>('tag');
+		assert.strictEqual((<any> RegistryWrapper).type, REGISTRY_ITEM);
+		const registryWrapper = new RegistryWrapper();
+		assert.strictEqual(registryWrapper.name, 'tag');
+		// These will always be undefined but show the type inference.
```
Comment (dylans) [inline tests/unit/tsx.ts]: "show the type inference" ... I'd be more explicit, e.g. "show the type inference of ___________"

**Comment by agubler** [comment]
@dylans comments addressed

**Comment by agubler** [comment]
Merging as the CI failures are unrelated functional test failures.

**Pull request merged**